packages/@aws-cdk/aws-eks/lib/cluster.ts

```diff
@@ -904,7 +1023,21 @@ export class Cluster extends Resource implements ICluster {
     let provider = this.stack.node.tryFindChild(uid) as KubectlProvider;
     if (!provider) {
       // create the provider.
-      provider = new KubectlProvider(this.stack, uid);
+
+      let providerProps: KubectlProviderProps = {};
+
+      if (!this.endpointAccess!.endpointPublicAccess) {
+        // endpoint access is private only, we need to attach the
+        // provider to the VPC so that it can access the cluster.
+        providerProps = {
+          ...providerProps,
+          vpc: this.vpc,
+          vpcSubnets: { subnetType: ec2.SubnetType.PRIVATE },
```

Comment:
Just curious - Does that mean it will just pick up all the private subnets and we would not be able to specify which subnets? 

The reason for asking this is that in our context, 1 VPC will have 10+ private subnets across 3 AZs. So if we don't have a choice to specify subnets as a comma separated string, then the corresponding Custom Resource Handler Lambda might want to bind to all those subnets and it would fail in our context due to Lambda cannot associate with multiple subnets in one AZ.

Reply:
@JasperW01 Good call out. 

We actually already accept a subnet selection for the VPC:

https://github.com/aws/aws-cdk/blob/e8c0b58cf3bf12fa6d9ca1bbc36c9fcee216098d/packages/%40aws-cdk/aws-eks/lib/cluster.ts#L130-L144

So I guess this should be:

```typescript
vpcSubnets: props.vpcSubent || { subnetType: ec2.SubnetType.PRIVATE },
```

Reply:
Hi @iliapolo , good pick up. 

I guess it will do the trick. But on the other hand, it would put an assumption that the user has only selected 1 subnet per AZ as the input parameter for cluster creation. However it may or may not the case. If not, then the CR Handler Lambda VPC association with subnets will still fail due to more than one subnet per AZ.

So I could think of two choices: 

Option i. Use "vpcSubnets: props.vpcSubent || { subnetType: ec2.SubnetType.PRIVATE }" as the fix here, but specifically document that - when creating cluster, please only choose 1 subnet per AZ for control plane and later you can still add data plane to other subnets in the same VPC anyway, as stated in the section "Sidebar: Do my worker nodes need to be in the same subnets as the ones I provided when I started the cluster?" of https://aws.amazon.com/blogs/containers/de-mystifying-cluster-networking-for-amazon-eks-worker-nodes/. 

(BTW, as a side point - for customer who has complex subnets in the VPC, this key point stated in this blog is quite important but not stated clearly in the EKS official document though. There is just a general link to this blog quoted in the EKS official document - https://docs.aws.amazon.com/eks/latest/userguide/create-public-private-vpc.html. I guess it might be worthwhile to call it out specifically in the cluster-creation sections of EKS official documents which will be very useful for customer who has many subnets in VPC rightly or wrongly. ) 

Option ii. Leave a specific option here to allow customer to specify subnets for the Custom Resource Handler Lambda to associate, so the customer has the chance to make sure only 1 subnet per AZ here. 

There might be other choices. By weighing simplicity & consistency with flexibility, my personal opinion would lean towards option i. But I guess to decide which one is better also needs to consider other factors of this aws-eks module holistically so I could be totally wrong to prefer option i. 

Anyway just my 2 cents for your reference.

Reply:
@JasperW01 Thanks again for the input. Let me ponder this a bit 👍

Reply:
Hi @JasperW01 - So you mentioned earlier:

> Handler Lambda might want to bind to all those subnets and it would fail in our context due to Lambda cannot associate with multiple subnets in one AZ

Are you sure about this? Im asking because I ran this scenario and saw that it does actually work. For example, this is my lambda VPC config:

<img width="1282" alt="Screen Shot 2020-07-22 at 5 28 00 PM" src="https://user-images.githubusercontent.com/1428812/88189294-956b5000-cc41-11ea-968f-85c341088628.png">

Notice that there are 3 subnets in each az, and the lambda binds to all of them. Am I missing something? Is there documentation somewhere about this multiple subnets thing?

Reply:
Hi @iliapolo , this is interesting and challenging my memory. I remember recently when I was provisioning Lambda with VPC association, it failed with an error message of "failed to associate multiple subnets per AZ". I will retrospect what happened and get back to you later today. 

If this Lambda subnet binding concern turns out to be a false alarm, then I guess "vpcSubnets: props.vpcSubent || { subnetType: ec2.SubnetType.PRIVATE }," would be all cool. In our context, we have around 20 subnets segregated to multiple management zones. So we need to specify the selected subnets by using "readonly vpcSubnets?: ec2.SubnetSelection[];" when creating EKS cluster, so it would be nice to align the Handler Lambda subnets with EKS control plane's subnets where the EKS control ENIs will sit.

Reply:
Hi @iliapolo I dug a bit deeper on my side and also double checked with my team mate. Both of us clearly remember we did encounter such error previously this year while using CDK to provision Lambda with VPC association. But funny thing is we could not replicate the exact error any more today so we suspect it might have been some CDK related issue and have been fixed along the way. 

On the other hand, today we did generate another issue when blanketly using ec2.SubnetType.PRIVATE to create Lambda in CDK codes in our env. The error is it exceeds 16 subnet limit as shown below: 

![image](https://user-images.githubusercontent.com/14971090/88244015-acf41880-ccd5-11ea-8a7b-e5800deef0e5.png)

so I guess it still makes sense to align the Handler Lambda subnet association with the EKS control plane subnet selection.